### PR TITLE
Use symbol tables for table definitions loaded from parquet files when appropriate.  Fixes #847

### LINKED ISSUE #847: Use symbol tables when is appropriate when reading parquet files.
After some discussions we decided that, for string columns, if every data page for a column uses dictionary encoding, we are safe to use a DH table definition with a symbol table.  Implement that.

### PR DESCRIPTION
**Pull request by jcferretti**
We check that every data page for the column uses dictionary encoding; known that we use a symbol table in the DH table definition.

### CODE REVIEW

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package io.deephaven.parquet;
 
 
+import io.deephaven.UncheckedDeephavenException;
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:4]: Unused

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java [hunk 2]
```diff
@@ -75,6 +77,56 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
         type = fromParquetSchema(fileMetaData.schema,fileMetaData.column_orders);
     }
 
+    private HashSet<String> columnsWithDictionaryUsedOnEveryDataPage = null;
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:80]: We usually don't overspecify return types/variable types, as convention in Java. Better for this to be `Set`, unless you need `HashSet`-specific methods or features.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java [hunk 3]
```diff
@@ -75,6 +77,56 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
...
+        if (columnsWithDictionaryUsedOnEveryDataPage == null) {
+            columnsWithDictionaryUsedOnEveryDataPage = calculateColumnsWithDictionaryUsedOnEveryDataPage();
+        }
+        return columnsWithDictionaryUsedOnEveryDataPage;
+    }
+
+    private HashSet<String> calculateColumnsWithDictionaryUsedOnEveryDataPage() {
+        final HashSet<String> result = new HashSet<>(fileMetaData.getSchemaSize());
+        int rowGroupSeq = 0;
+        for (RowGroup rowGroup : fileMetaData.getRow_groups()) {
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:91]: What kind of monster wrote this library? Camel case mixed with underscores? ;-)

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java [hunk 4]
```diff
@@ -75,6 +77,56 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
...
+            // that satisfy the restriction.
+            // On later passes after zero, we will remove any column that does not satisfy
+            // the restriction.
+            for (ColumnChunk columnChunk : rowGroup.columns) {
+                final ColumnMetaData columnMeta = columnChunk.getMeta_data();
+                final String parquetColumnName = columnMeta.path_in_schema.get(0);
+                if (rowGroupSeq > 0 && !result.contains(parquetColumnName)) {
+                    continue;
+                }
+                if (columnMeta.encoding_stats == null) {
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:102]: What does this mean? Is it possible to have stats on the first column chunk, but not on subsequent ones? Seems like we're assuming that if that happens, it doesn't create a conflict. I might rather assume that it does, and remove this column from the set if we see no stats.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java [hunk 5]
```diff
@@ -75,6 +77,56 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
...
+                    continue;
+                }
+                boolean thisColumnInThisRowGroupDisproved = false;
+                for (PageEncodingStats encodingStat : columnMeta.encoding_stats) {
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:106]: So, we think this is a relatively cheap/low-cost thing to do? Do we always read the entire metadata in one go, or are we in fact causing file accesses here? I think this approach is predicated on it being "cheap enough". 

Alternatively, we could pull this idea out of column defs, move it to the column source level, and have column sources do this check when asked questions about symbol table support.

**Review by rcaudy — COMMENTED** [review]
I have a few trivial comments and two sort-of important questions in the issues.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java (hunk 3 above)
Comment (jcferretti) [reply Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:91]: Is code generation, for thrift.   Click on the method or type and take a peek at the file you land on.  Don't do it on a full stomach.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java (hunk 4 above)
Comment (jcferretti) [reply Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:102]: The logic was mudded.  Is easier to reason about if we just explicitly separate the first pass.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java (hunk 5 above)
Comment (jcferretti) [reply Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:106]: This stuff is still in the thrift file metadata object; if we had to read it from the page preamble, that would mean seeking to the page positions, this is not doing that.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java [hunk 6]
```diff
@@ -77,52 +73,66 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
...
-        int rowGroupSeq = 0;
-        for (RowGroup rowGroup : fileMetaData.getRow_groups()) {
-            // On the first pass, for row group zero, we are going to add all columns to the set
-            // that satisfy the restriction.
-            // On later passes after zero, we will remove any column that does not satisfy
-            // the restriction.
+        final List<RowGroup> rowGroups = fileMetaData.getRow_groups();
+        final Iterator<RowGroup> riter = rowGroups.iterator();
+        if (!riter.hasNext()) {
+            return result;
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:111]: I think this should be the full map, not the empty map, maybe? Do we have a way to get such columns, without a row group? This is probably a meaningless distinction in the single partition case.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java [hunk 7]
```diff
@@ -77,52 +73,66 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
...
+            if (encodingStat.encoding != Encoding.PLAIN_DICTIONARY
+                    && encodingStat.encoding != Encoding.RLE_DICTIONARY) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Set<String> calculateColumnsWithDictionaryUsedOnEveryDataPage() {
         final HashSet<String> result = new HashSet<>(fileMetaData.getSchemaSize());
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:107]: `HashSet`->`Set`, but only if you have to edit this file for another reason.

**Review by rcaudy — APPROVED** [review]

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java (hunk 6 above)
Comment (jcferretti) [reply Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:111]: We do have a way to get all column names without going through the chunks, we can just do:
```
            for (SchemaElement se : fileMetaData.getSchema()) {
                if (!se.isSetNum_children()) {  // We want only the leaves.
                    result.add(se.getName());
                }
            }
```
(I was thinking of doing it and wrote the code above).

But I am changing my mind.  I think if we define the property as:

```
    /**
     * Get the name of all columns that we can know for certain
     * (a) have a dictionary, and (b) use the dictionary on all data pages.
     *
     * @return A set of parquet column names that satisfies the required condition.
     */
    public Set<String> getColumnsWithDictionaryUsedOnEveryDataPage() {
    ...
```

This is properly empty, and I think that is a better result in terms of behavior (goes with the "when we can't tell")

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java (hunk 7 above)
Comment (jcferretti) [reply Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java]: Done.

File: Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java (hunk 6 above)
Comment (jcferretti) [reply Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java:123]: Changed my mind, you convinced me, now we return all columns.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**